### PR TITLE
Nominal Types

### PR DESCRIPTION
**Pull request by asutula**

### CODE REVIEW

**Comment by asutula** [comment]
This is good to go @andrewxhill

File: App/Components/ProgressiveImage.tsx [hunk 1]
```diff
@@ -2,9 +2,10 @@ import React, { Fragment } from 'react'
 import { View } from 'react-native'
 import { ImageStyle, Platform } from 'react-native'
 import TextileImage from '../../TextileImage'
+import { PhotoId } from '../Models/TextileTypes'
 
 export interface IProgressiveImageProps {
-  imageId: string,
+  imageId: PhotoId,
```
Comment (asutula) [inline App/Components/ProgressiveImage.tsx]: Decided that UI components that really require a certain id type should just be typed strongly.

Comment (asutula) [reply App/Components/ProgressiveImage.tsx]: In contrast, if we create some general purpose/reusable UI component, it's props should just be normal types like `string` or `number`, for example.

File: App/Services/Notifications.ts [hunk 2]
```diff
@@ -9,23 +9,65 @@ export interface INotificationsPayload {
   typeString: string,
 }
 
-export function isPhoto(notification: Notification): boolean {
-  if (!notification.data_id) {
-    return false
-  }
-  switch (notification.type) {
-    case NotificationType.receivedInviteNotification:
+export function toTypedNotification(notificationData: NotificationData): Notification {
```
Comment (asutula) [inline App/Services/Notifications.ts]: This is the function that converts the untyped `NotificationData` to type safe notifications.

File: App/Services/Notifications.ts [hunk 3]
```diff
@@ -9,23 +9,65 @@ export interface INotificationsPayload {
...
+  const { subject_id, subject, data_id, type, ...baseNotification } = notificationData
+  switch (type) {
+    case NotificationType.commentAddedNotification:
+      return {
+        ...baseNotification,
+        type,
+        threadId: notificationData.subject_id as any,
```
Comment (asutula) [inline App/Services/Notifications.ts]: Any time you have a trusted normal value (`string` in this case) that you need to convert to the nominal type, you use `as any`.

File: App/Models/TextileTypes.ts [hunk 4]
```diff
@@ -5,35 +5,21 @@
...
+export type DeviceId = string & { _deviceIdBrand: void }
+export type ProfileAvatarId = string & { _profileAvatarIdBrand: void }
+export type NotificationId = string & { _notificationIdBrand: void }
+
+export type UserName = string & { _userNameBrand: void }
+export type ThreadName = string & { _threadNameBrand: void }
+export type DeviceName = string & { _deviceNameBrand: void }
+export type PublicKey = string & { _publicKeyBrand: void }
+export type PrivateKey = string & { _privateKeyBrand: void }
+export type Mnemonic = string & { _mnemonicBrand: void }
```
[the remark this answers is not in the document]
Comment (andrewxhill) [reply App/Models/TextileTypes.ts]: nice

**Pull request merged**